utf8/utf8.go

```diff
@@ -0,0 +1,9 @@
+package utf8
+
+import _ "github.com/segmentio/asm/cpu"
+
+// Valid reports whether p consists entirely of valid UTF-8-encoded runes.
+func Valid(p []byte) bool {
+	ret, _ := Validate(p)
+	return ret
+}
```

Comment:
I wonder if we could move the `ascii.Valid` check to this function, then rename `utf8.Validate` to something like `utf8.ValidNotASCII`, and we can get rid of the double boolean return value. What do you think?

```suggestion
// Valid reports whether p consists entirely of valid UTF-8-encoded runes.
func Valid(p []byte) bool {
	if ascii.Valid(p) {
		return true
	}
	return ValidNotASCII(p)
}
```

The compiler should be able to inline the `utf8.Valid` function as well here and remove one function call in the common case where the strings contain only ASCII characters.

Reply:
I think it depends on what inputs we want to optimize for. When I tried the way you described, always calling `ascii.Valid` before running the UTF8 AVX validation function added some notable overhead. It probably comes from one more stack preparation, and creates more work when the payload is mostly ASCII but not towards the end. I think we could mitigate both problems by inlining `ascii.Valid` (or figuring out how to use the register based calling convention?) and having it return the offset of the non-ASCII byte it found so that the UTF8 validation can start from there.

Reply:
Thinking a bit more about it from an API perspective: if we want to write code that has different optimized paths depending on if the input is ASCII-valid or UTF-8-valid, we would have to run the validation twice to know, or re-implement this function directly.

Reply:
> **(or figuring out how to use the register based calling convention?)**

There is no way to do this in Go at this time, the compiler will always use the stack to pass arguments to assembly functions.

> having it return the offset of the non-ASCII byte it found so that the UTF8 validation can start from there

This seems like it could be useful but maybe outside the scope of this PR?

I see how we're only calling `ascii.Valid` for short strings (I assumed we called it every time). In that case, I don't know if we need the `Validate` function at all, it seems like it would never tell for sure whether something is ASCII (a long valid UTF-8 string could be ASCII but it would say it's not).

Should we just remove the `Validate` function then?

Reply:
> There is no way to do this in Go at this time, the compiler will always use the stack to pass arguments to assembly functions.

I meant have our assembly function to call `ascii.Valid` the same way the compiler would generate code to use the new register-based ABI.

> (a long valid UTF-8 string could be ASCII but it would say it's not

Not sure I understand. Barring any bug, `validateAvx` should only `validAscii` as `true` iff every single character in the input is in the ASCII range.

IIRC you were thinking of some parsing optimizations in encoding/json to have 3 code paths depending on whether a string is ASCII-only, UTF-8-valid, or not valid UTF-8. Maybe I misunderstood that?

Reply:
> I meant have our assembly function to call ascii.Valid the same way the compiler would generate code to use the new register-based ABI.

I don't think that's possible either, the assembly code will be compiled to read arguments from the stack. We can always just assume the calling convention and read arguments from registers but this becomes a bit brutal in terms of maintenance.

---

Sorry, by "bugged" I meant "bothered", I didn't mean that the logic was wrong.

What I'm concerned about is the caller inverting the results:
```go
utf, ascii := utf8.Validate(x)
```
```go
ascii, utf := utf8.Valiate(x)
```

Both of these pass compilation just fine, and could be really hard to detect. This is the type of bug that would likely make it to production, nobody will write a test that actually uncovers the misuse of the API.

I would like us to ship an API that is harder to misuse, which is why I'm suggesting we use a bit set to represent the tri-state, then usage would look like this:
```go
if utf8.Validate(x) != utf8.Invalid {
  ...
}
```
```go
switch state := utf8.Validate(x); {
case (state & utf8.ASCII) != 0:
  ...
case (state & utf8.UTF) != 0:
  ...
default:
  ...
}
```
We can make the API even safer with a type that helps use the bit set properly:
```go
type Validation uint

func (v Validation) IsASCII() bool { return (v & ASCII) != 0 }

func (v Validation) IsUTF() bool { return (v & UTF) != 0 }

func (v Validation) IsInvalid() bool { return v == Invalid }

func Validate(p []byte) Validation { ... }
```

Reply:
I think you added idea of the bitset after I answered :D Okay then, I completely agree with using a custom type to represent the three states. I thought your concern was about surfacing the is-ASCII information at all.

I'll make the change to the API.